<a href="https://colab.research.google.com/github/ilovesylus/mathcomputer/blob/main/Quiz1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Computer Programming Quiz 1

#### Student ID: B132045005

#### Name: 李子玉

## A.	Single Choice Questions (4%, 48%)

### (1) Which of the following expressions, when evaluated in a Python control flow condition, returns a Boolean value that differs from the others?

(A) `[]`

(B) `None`

(C) `"None"`

(D) `2 > 5 or 4 < 3`

> Ans: C

### (2) Which of the following statements is false?

(A) Python’s built-in function `type()` returns the data type of an object.

(B) In a function call, the parentheses following the function name enclose the function’s arguments — that is, the data the function requires to perform its task.

(C) Python is case sensitive; therefore, the identifiers `pokemon` and `Pokemon` may refer to different objects.

(D) A function is executed by writing its name followed by curly braces `{}`.

> Ans: D

### (3) Which of the following statements is true?

(A) Python uses the `^` operator to raise one value to the power of another (e.g., `2 ^ 10`).  

(B) The value of the expression `7.5 // 3` is 2.5.  

(C) The literal `2e5` represents `32.0` in Python.

(D) Python uses indentation to define code blocks.

> Ans: D

### (4) Which of the following statements is False regarding Python control flow?

(A) All statements within a loop body must be indented to the same level.

(B) A while loop is recommended when you know in advance how many times you want the loop to execute.

(C) Both `continue` and `break` can be used to control iterations in `for` and `while` loops.

(D) Calling `sys.exit()` will terminate the program immediately, regardless of pending instructions.

> Ans: B

### (5) Which of the following statements about sets and dictionaries in Python is True?

(A) Sets are ordered collections, so you can access elements by index using square brackets.

(B) Dictionary keys can be any type, including mutable types like lists.

(C) The `set()` function returns an empty set, whereas `{}` creates an empty dictionary.

(D) Both sets and dictionaries support indexing and slicing using square brackets.

> Ans: C

### (6) Which of the following statements is true?

(A) The `random.randint(1, 5)` function call will randomly generate integers from 1 to 4.

(B) The expression `list(range(1, 5))` returns the list `[1, 2, 3, 4, 5]`.

(C) If a function has no `return` statement, it implicitly returns `None` after executing its last statement.

(D) In function definitions, parameters with default values must always be placed on the leftmost side of the parameter list.

> Ans: C

### (7) What does the following line of code display?

```python
print(20, 30, 40, sep='; ')
```

(A) 203040

(B) 20;30;40

(C) 20; 30; 40

(D) 20 30 40

> Ans: C

### (8) Which of the following statements is false?

(A) A local variable is accessible only within the function that defines it.

(B) A function can access a global variable's value without any special declaration.

(C) To modify a global variable inside a function, you must declare it with the `global` keyword.

(D) Using the asterisk (`*`) in an `import` statement imports every function from a module, and this practice is recommended for larger modules.

> Ans: D

### (9) Which of the following is not a correct way to create a string in `Python`?

(A) `"nsysu"`

(B) `'nsysu'`

(C) `""" nsysu """`

(D) `str(nsysu)`

> Ans: D

### (10) Assume we have a list `x = ['m','a','t','h']`, what is the result of `x[::-1]`?

(A) `['m','a','t','h']`

(B) `'h'`

(C) `['m','a','t']`

(D) `['h','t','a','m']`

> Ans: D

### (11) Assume the following assignment statements are executed in Python, and the user inputs an integer:
```python
spam1 = input()
spam2 = int(spam1)/5
spam3 = spam2 + 18

```
What are the data types of the variables `spam1`, `spam2`, and `spam3`, respectively?

(A) int, int, int   

(B) str, int, int

(C) str, float, float   

(D) str, float, int

> Ans: C

### (12) Which of the following statements is false?

(A) Dividing by zero results in a runtime error.

(B) Omitting a closing parenthesis when calling a function produces a syntax error.

(C) Misspelling a function name is considered a semantic error.

(D) Misunderstanding the order of operations when writing code constitutes a semantic error.

> Ans: C

## B.	Short-answer and coding questions

### (13)  Write a program that prompts the user to enter a distance in centimeters (assume the input is a positive integer of at least 100,000). Convert this distance into kilometers, meters, and remaining centimeters. Print the result in the following format: (10%)

```
<kilometers> km - <meters> m - <centimeters> cm
```
For example, if the user inputs 123456 centimeters, the program should calculate and print:

```
1 km - 234 m - 56 cm
```

Hint: Use the floor division and the remainder operation to calculate the number of hours, minutes, and seconds.

In [ ]:
# Your code here

1 km - 234 m - 56 cm


In [ ]:
centimeters = int(input("Enter a distance in centimeters (at least 100,000): "))

kilometers = centimeters // 100000
meters = (centimeters % 100000) // 100
remaining_centimeters = centimeters % 100

print(f"{kilometers} km - {meters} m - {remaining_centimeters} cm")

Enter a distance in centimeters (at least 100,000): 123456
1 km - 234 m - 56 cm


### (14)   Write programs that display the following triangle patterns using `for` loops. (10%)

Hint: You can display a hash symbol using:

```python
print('#', end='')
````

(a)

```
    #
   ###
  #####
 #######
#########
```

(b)

```
#########
 #######
  #####
   ###
    #
```

In [ ]:
# Your code here

rows = 5

for i in range(1, rows + 1):
    print(' ' * (rows - i), end='')
    print('#' * (2 * i - 1))


    #
   ###
  #####
 #######
#########


In [ ]:
# Your code here
rows = 5

for i in range(rows, 0, -1):
    print(' ' * (rows - i), end='')
    print('#' * (2 * i - 1))


#########
 #######
  #####
   ###
    #


### (15) Write a program that computes the value of π using the Nilakantha series. The series is given by (10%):

$$ \pi = 3 + \sum_{n=1}^{\infty} (-1)^{n+1} \cdot \frac{4}{(2n)(2n+1)(2n+2)} $$

Your task is to write a program that keeps adding terms from the series until the absolute error (i.e., the difference between your computed value and `math.pi`) is less than $10^{-3}$. Finally, print the computed value of π along with the number of terms used, separated by a dash. For example, the output might look like:

```
3.1416 - 2000
```

*Hint:*  
- Use `math.pi` from the Python `math` module as the ground truth value.  
- Use the `abs()` function to calculate the error.  
- The first term of the series (for $ n = 1 $) is $ \frac{4}{2 \times 3 \times 4} $.

In [ ]:
import math

# Your code here

In [ ]:
import math

def compute_pi():
    # 初始值，Nilakantha 公式的起始值是 3
    pi_estimate = 3
    n = 1
    terms_used = 0

    # 一直迭代直到誤差小於 10^-3
    while True:
        # 計算 Nilakantha 公式中的當前項目
        term = 4 / (2 * n * (2 * n + 1) * (2 * n + 2))

        # 如果 n 是奇數，則加上當前項，偶數則減去
        if n % 2 == 1:
            pi_estimate += term
        else:
            pi_estimate -= term

        # 計算當前誤差
        terms_used += 1

        # 如果誤差小於 10^-3，則停止
        if abs(pi_estimate - math.pi) < 10**-3:
            break

        # 增加 n 來進行下一輪計算
        n += 1

    # 輸出最終計算結果，精確到小數點後 4 位
    print(f"{pi_estimate:.4f} - {terms_used}")

# 呼叫函數
compute_pi()


3.1409 - 6


> Ans: 3.1409 - 6

We need 6 terms in order to get the error below $10^{-3}$

### (16) Here, we will simulate the process of a simple card game. The game is played with a standard deck of 52 cards, and we will randomly select 20 cards and divide them evenly between two players. Each player gets a hand of 10 cards. The goal of the game is to determine whether the player got a flush (i.e., at least five cards of the same suit) or not in his 10 cards. If any suit appears five or more times, please return `True`. Otherwise, return `False`. Please complete the following game design by filling in the missing part in the function. (12%).

In [ ]:
import random

RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]

def create_deck():
    suits = ['♣', '♦', '♥', '♠']
    deck = [(rank, suit) for rank in RANKS for suit in suits]  # Use list comprehension to create the deck.
    return deck

# This function shuffles the deck, selects 20 cards, and deals two hands of 10 cards each.
def deal_cards(deck):
    random.shuffle(deck)
    deck = deck[:20]
    hand1 = deck[:10]  # Split into 10 cards for each hand using slicing.
    hand2 = deck[10:]
    return hand1, hand2

def is_flush(hand):
    """Return True if hand contains a flush (at least 5 cards of the same suit); otherwise, return False."""
    # Your code here
    suit_count = {}  # Dictionary to count suits
    for card in hand:
        suit = card[1]  # Extract the suit from the card (second element of the tuple)
        if suit in suit_count:
            suit_count[suit] += 1
        else:
            suit_count[suit] = 1

    # Check if any suit appears 5 or more times
    for count in suit_count.values():
        if count >= 5:
            return True

    return False

In [ ]:
# Test data 1
random.seed(2025)
deck = create_deck()
hand1, hand2 = deal_cards(deck)
f1 = is_flush(hand1)
f2 = is_flush(hand2)
print("Test data 1:", f1, f2)

# Test data 2
random.seed(20)
deck = create_deck()
hand1, hand2 = deal_cards(deck)
f3 = is_flush(hand1)
f4 = is_flush(hand2)
print("Test data 2:", f3, f4)


assert (f1,f2) == (True, False)
assert (f3,f4) == (False, False)

Test data 1: True False
Test data 2: False False


### (17) Considering the following code snippet: (10%)

```python
def add_event(calendar):
    calendar['Friday'].append('Conference')
    print("Inside function:", calendar)

weekly_calendar = {'Monday': ['Meeting'], 'Friday': ['Lunch'], 'Saturday': ['Hike']}
calendar_copy = weekly_calendar  # Modify this line so that changes inside add_event do not affect weekly_calendar.
add_event(calendar_copy)
print("Outside function:", weekly_calendar)
```

(a) What will be printed when executing this code, and why does this happen?

(b) How can you modify the assignment of `calendar_copy` so that the changes made inside `add_event` do not alter the original `weekly_calendar`?

> Ans: *double click here to answer the question.*